### PR TITLE
Add an adaptation of Alex Shinn’s match-simple

### PR DESCRIPTION
**Pull request by aartaka**
Hi! This adds a version of [match-simple.scm](https://synthcode.com/scheme/match-simple.scm) rewritten in non-hygienic `define-macro`-s. Feature set should be roughly the same, except that it also allows non-quoted constants (chars, strings, numbers, booleans,) where Shinn’s matcher would require quoting them.

One significant problem is that `_` pattern is breaking the macro. @egallesio any way to allow `_` as an aux syntax?

How does it look in general?

### CODE REVIEW

**Comment by jpellegrini** [comment]
Already @aartaka ? You mentioned only two days ago that you'd  work on it... :smile:

@egallesio look at this:

```scheme
(pp (macro-expand '(match (list 1 2 3 4) ((a b c d) (+ a b c d)))))

(let ((expr2562 (list 1 2 3 4)))
  (cond ((and (pair? expr2562)
              (let ((car2563 (car expr2562))
                    (cdr2564 (cdr expr2562)))
                (and #t
                     (and (pair? cdr2564)
                          (let ((car2565 (car cdr2564))
                                (cdr2566 (cdr cdr2564)))
                            (and #t
                                 (and (pair? cdr2566)
                                      (let ((car2567 (car cdr2566))
                                            (cdr2568 (cdr cdr2566)))
                                        (and #t
                                             (and (pair? cdr2568)
                                                  (let ((car2569 (car cdr2568))
                                                        (cdr2570
                                                          (cdr cdr2568)))
                                                    (and #t
                                                         (null? cdr2570)))))))))))))
         (let* ((car2571 (car expr2562))
                (a car2571)
                (cdr2572 (cdr expr2562))
                (car2573 (car cdr2572))
                (b car2573)
                (cdr2574 (cdr cdr2572))
                (car2575 (car cdr2574))
                (c car2575)
                (cdr2576 (cdr cdr2574))
                (car2577 (car cdr2576))
                (d car2577)
                (cdr2578 (cdr cdr2576)))
           (+ a b c d)))))
```

Seems quite short, compared to the one you showed in #935

**Comment by aartaka** [comment]
@jpellegrini I mean, I can shorten it even further, but I decided that a minimal implementation is more reviewable than an optimized one. To lessen the load on @egallesio :wink:

**Comment by aartaka** [comment]
One thing that isn’t there in this PR is record destructuring. It’s definitely possible to do, and Bigloo match has it. But, again, let’s merge this first and extend later.

**Comment by egallesio** [comment]
Thank you very much @aartaka for this contribution. It looks great.  
I can't wait to test it and integrate it into STklos.

**Comment by egallesio** [comment]
> One significant problem is that `_` pattern is breaking the macro. @egallesio any way to allow `_` as an aux syntax?

I'll take a look at it.

**Comment by aartaka** [comment]
>  One significant problem is that _ pattern is breaking the macro. @egallesio any way to allow _ as an aux syntax?
>
> I'll take a look at it.

Oh, no, sorry, that was silly me not importing (smatch). I’ve fixed it
and now work with broken tests! Will ping you once it all passes!

**Comment by aartaka** [comment]
Okay I’ve fixed tests. I commented out one of them because supporting it
would take a lot of work for a really marginal corner-case. @egallesio,
this is ready for review.

**Comment by aartaka** [comment]
> Okay I’ve fixed tests. I commented out one of them because supporting it
> would take a lot of work for a really marginal corner-case. @egallesio,
> this is ready for review.

To be fair, match-simple.scm does not pass this test either.

File: lib/smatch.stk [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+      `(let ([,expr-var ,expr])
+         (cond
+          ,@(map
+             (lambda [clause]
+               (match-case
+                clause
+                [(?patt . ?body)
+                 `(,(clause->condition expr-var patt)
+                   (let* (,@(clause->bindings expr-var patt))
+                     ,@body))]))
```
Comment (egallesio) [inline lib/smatch.stk]: We should add a systematic else clause such as
 (else (error 'match "no matching clause for ~S" ,expr-var))

To yield an error when we have no matching clause

File: lib/smatch.stk [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+           ,@clauses))))
+
+  (define-macro (match-let . body)
+    (if (symbol? (car body))
+        (let* ([loop-name (car body)]
+               [bindings (cadr body)]
+               [vars (map (lambda _ (gensym "match-let-var")) bindings)]
+               [body (cddr body)])
+          `(let ,loop-name (,@(map (lambda [var binding]
+                                     (list var (cadr binding)))
```
Comment (egallesio) [inline lib/smatch.stk]: should be binding[s]

**Comment by egallesio** [comment]
@aartaka, some informations on this PR:
I'm still working on this PR. I have fixed little issues, but mainly, I rewrote it to not use `match-case` (the idea, is to replace it by your code, when it will be finalized). The remaining problem, I would like to solve is that 

```scheme
match '(1 2) ((a a) a) (_ 'no))
```
 returns 2 instead of `no`. 
I know that `match-simple.scm` does not manage this case, but it would be cool, if I could fix it.

**Comment by aartaka** [comment]
> @aartaka, some informations on this PR:
> I'm still working on this PR. I have fixed little issues, but mainly, I rewrote it to not use match-case (the idea, is to replace it by your code, when it
> will be finalized).

Thanks for keeping me updated! 

> The remaining problem, I would like to solve is that
>
> match '(1 2) ((a a) a) (_ 'no))
>
> returns 2 instead of no.
> I know that match-simple.scm does not manage this case, but it would be cool, if I could fix it.

Wow, that’s a really hard problem! I didn’t dare to approach it,
sticking to match-simple feature set for implementability. But if you’re
feeling adventurous, then sure, your call!
